### PR TITLE
Data transformation

### PR DESCRIPTION
**Pull request by dblatcher**
## What does this change?

Outlines a system for transforming our TBC data model to the legacy API format in the data-client library.
 - a new data model can be defined and edited as a `zodSchema` and the type and validation functions generated from the schema
 - The `deriveNewsletter` function has to be updated to take account of any changes to the `newsletterDataSchema`
 - A set of unit test can verify that the transformation module works as expected.

## How to test
`nx run newsletters-data-client:test --testFile=transform` will run the unit tests (they should pass)

Making changes to `newsletterDataSchema`, such as removing a field  required on the `Newsletter` type should make the test fail, until the `deriveNewsletter` function is updated to ensure the field is added as required.

E.G.  try removing  `listIdV1: z.number(),` from the `newsletterDataSchema` and the test will fail

## How can we measure success?

It should be possible to define our new data model using a `zodSchema` and convert it safely to the legacy API format. 

*update* - the advantage of having this in the codebase before we know what our data model will be is that it should allow us to make sure that it remains possible to derive the old data from the new data model as we develop it - for example, if we forget that we need to keep the `listId` field for backwards compatability, we should get a test error.

## Have we considered potential risks?

`newsletterDataSchema` is just a placeholder for the new data model. There are comments to indicate this and the type and functions are not exported by the data client library.

We will need to expand on the unit test as we go to ensure that the converted data has the correct values (not just the correct type).

### CODE REVIEW

**Comment by codingWithRach** [comment]
I feel that we might be heading for a terminology issue.  We have a file newsletter-type.ts (containing baseNewsletterSchema and two extensions one of which is called newsletterSchema) and we have a file newsletter-data-type.ts (containing newsletterDataSchema).  It's not immediately clear what the difference is between a -type and a -data-type; and between a newsletterSchema and a newsletterDataSchema.

File: libs/newsletters-data-client/src/lib/schema-helpers.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+export const emailEmbedSchema = z.object({
+	name: z.string(),
+	title: z.string(),
+	description: z.string().optional(),
+	successHeadline: z.string(),
+	successDescription: z.string(),
+	hexCode: z.string(),
+	imageUrl: z.string().optional(),
+});
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply libs/newsletters-data-client/src/lib/schema-helpers.ts]: 👍

File: libs/newsletters-data-client/src/lib/transform.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	try {
+		const merged: Newsletter & Partial<NewsletterData> = {
+			...newsletterData,
+			...deriveBooleansFromStatus(newsletterData.status),
+		};
+
+		delete merged.creationTimeStamp;
+		delete merged.status;
+
+		return merged;
```
[the remark this answers is not in the document]
Comment (dblatcher) [reply libs/newsletters-data-client/src/lib/transform.ts]: 👏  nice - I've done that and added an eslint ignore to get rid of the 'unused vars' warning.

**Comment by dblatcher** [comment]
> Fab PR! Great description, really neat code and tests making it easy to understand.
> 
> Agree with Rachel that the naming is getting a bit confusing! Could do a follow up PR with name changes if you fancy.

Yes, makes sense! I think the distinction we want to capture is between the legacy `Newsletter` type that matches the API response and the data model for how we actually want to store newsletters. Maybe we should rename `Newsletter` to `NewsletterResponse` to match the convention in frontend?

**Comment by codingWithRach** [comment]
> > Fab PR! Great description, really neat code and tests making it easy to understand.
> > Agree with Rachel that the naming is getting a bit confusing! Could do a follow up PR with name changes if you fancy.
> 
> Yes, makes sense! I think the distinction we want to capture is between the legacy `Newsletter` type that matches the API response and the data model for how we actually want to store newsletters. Maybe we should rename `Newsletter` to `NewsletterResponse` to match the convention in frontend?

Or if something represents the legacy version, we could call it Legacyxxx, so that everybody knows at a glance exactly what it is?

**Comment by dblatcher** [comment]
I'll put a ticket on the board about renaming type

**Pull request merged**